ntoskrnl/mm/balance.c

```diff
@@ -293,7 +293,8 @@ MmRebalanceMemoryConsumers(VOID)
 {
     // if (InterlockedCompareExchange(&PageOutThreadActive, 0, 1) == 0)
     {
-        KeSetEvent(&MiBalancerEvent, IO_NO_INCREMENT, FALSE);
+        /* Reset the event and boost balancer priority */
+        KeSetEvent(&MiBalancerEvent, EVENT_INCREMENT, FALSE);
```

Comment:
I have done much testing with this PR and what I found is that this change does not improve the outcome. The commented line just above appears to have been an attempt to limit the times that this Event can be consecutively reset. I added a DPRINT1 to balance.c at the entry of MmRebalanceMemoryConsumers and you can see the effect in the debug log here;
```
(ntoskrnl/mm/ARM3/pfnlist.c:126) Running low on pages: 80 remaining
(ntoskrnl/mm/balance.c:301) MmRebalanceMemoryConsumers.
(ntoskrnl/mm/ARM3/pfnlist.c:126) Running low on pages: 79 remaining
(ntoskrnl/mm/balance.c:301) MmRebalanceMemoryConsumers.
(ntoskrnl/mm/ARM3/pfnlist.c:126) Running low on pages: 78 remaining
(ntoskrnl/mm/balance.c:301) MmRebalanceMemoryConsumers.
(ntoskrnl/mm/ARM3/pfnlist.c:126) Running low on pages: 77 remaining
(ntoskrnl/mm/balance.c:301) MmRebalanceMemoryConsumers.
(ntoskrnl/mm/ARM3/pfnlist.c:126) Running low on pages: 76 remaining
(ntoskrnl/mm/balance.c:301) MmRebalanceMemoryConsumers.
. . .
```
Since pfnlist.c is a high priority function, it consistently hammers the KeSetEvent function resulting in multiple resets with no good outcome.

My suggestion is to uncomment and fix line # 294 above by changing it to the following:
```
if (InterlockedCompareExchange(&PageOutThreadActive, 1, 0) == 0)
```
After this fix, the Event can work as expected without being constantly reset.

Reply:
The way `PageOutThreadActive` is currently used is not technically safe -- if you uncomment that line, it may be that the balancer thread misses a wake-up. The set-event-if-interlocked pattern usually requires a lock in order to work right.
That may not be important for the balancer, given that it runs on a timer anyway, but https://github.com/reactos/reactos/commit/6f2b94c0231cbb447bc176712d11090b034490ee seems to indicate that this change was helping somehow.

Anyway, I'd prefer to separate this out from the cache trimming, and have more focused discussion/testing around this if someone strongly feels that either of the two changes is necessary.


(On a separate note, this boolean is pointless -- we could just make the event a `NotificationEvent` to achieve the same effect)

For posterity, the exact race condition around `PageOutThreadActive` looks like this:
* the balancer thread is inactive (`PageOutThreadActive=0`) and somebody calls `MmRebalanceMemoryConsumers`
* the `InterlockedCompareExchange` goes through, it sets `PageOutThreadActive` to 1 and signals the event
* `MiBalancerThread` wakes up and runs its loop. It finishes the loop but gets unscheduled before it decrements `PageOutThreadActive`
* another thread calls `MmRebalanceMemoryConsumers`. `PageOutThreadActive` is still 1 so it does not set the event
* `MiBalancerThread` decreases `PageOutThreadActive` and waits
* `MiBalancerEvent` did not get set so the thread remains until the next wake-up.

Reply:
I just wanted to point out that my suggested changed was not just to remove the comment marker.
Also, this line needs to have the '1' changed to a '0' and the '0' changed to a '1' in the InterlockedCompareExchange.
I modified the comment above to show the Previous line with a '-' followed by the New line with a '+'.
Hopefully this makes the changes more obvious. Just removing the commenting does not fix its intent.
It is fine to leave this line as it is for now. It can be revisited at a later time. Thanks.

Reply:
Thanks! Right, with `0, 1` the event would never get set... so that explains why Jérôme had to remove it